### ISSUE TITLE
How to update embadding score for Universal sentence encoder

### ISSUE THREAD

**Issue by SyedAgha** [tensorflow/hub#600]
Hi,

Currently I am using USE to find semantic similarity between differebt texts. Now I want to update the embedding scores without adding new vocabulary. I am working on medical documents which have some specific vocabulary and many keywords are used in a particular context. The documents are in pdf files and have no labels. I want to train USE on new documents to update the embedding score for keywords which in turn can help me improve the sementic similarity score. I tried to find any relevent document to explain this process but so far no success. 

How can I solve this problem, any help please?

Thanks!

**Comment by djdongjin** [comment]
Hi, 

If I understand correctly, you want to fine-tune the USE encoder with your own training data and labels. If so, I think you can do something like this:

```python
module_handler = "https://tfhub.dev/google/universal-sentence-encoder/4"
use_layer = hub.KerasLayer(module_handler, trainable=True)
model = tf.keras.Sequential([
     tf.keras.layers.InputLayer(input_shape=(...)),
     use_layer,
     tf.keras.layers.Dense(1)
])
```

Then you compile the model with MSE Loss and train with your data. You can find a similar workflow in this [tutorial](https://www.tensorflow.org/hub/tutorials/tf2_image_retraining) except that it is for image retraining.

**Comment by djdongjin** [comment]
Oh I just saw that your data has no labels. Then maybe you need a complete USE model training and start with the checkpoint provided by TF Hub, or other techniques such as negative sampling. I am not sure :)

**Comment by SyedAgha** [comment]
Thanks for reply.  Actually I have text files but I have no labels. I would like to update the sentence embedding score based on domain specific data. Do you have any idea how can I go further?

**Comment by djdongjin** [comment]
I see your point. I think unsupervised fine-tuning is similar to the pretraining process itself, and can be done by the embedding framework itself (e.g. USE, word2vec). So maybe you can have a look on the USE paper to see how they construct training examples and what losses they use. Then you can construct your dataset similarly with your own text. 

Notice that you don't need to implement the USE model itself, because it is already a complete encoder. The only thing you need to add might be an output layer to compute the loss.

**Comment by SyedAgha** [comment]
Thanks for your reply. Actually I am new in this field. Can you explain the last part of your reply please?

"Notice that you don't need to implement the USE model itself, because it is already a complete encoder. The only thing you need to add might be an output layer to compute the loss."

**Comment by djdongjin** [comment]
For instance, if you want to train a USE model, you need to implement the encoder model, which is a transformer here, and one or more output layers to compute losses (e.g. a binary cross-entropy loss). Then you train the complete model by constructing training data from text.

With the checkpoint provided by TF Hub (e.g. hub.KerasLayer(...)), you don't need to implement the encoder itself because this is what the magic TH Hub link provides. Then you can start with the checkpoint and add the corresponding output layer and fine-tune the model on your own domain-specific data following the way how they construct training data and compute the loss. 

This is just the way that I think is viable to unsupervised fine-tune a text encoder.

**Comment by SyedAgha** [comment]
Thanks a lot, you have been really helpful. A last question:  in hub.Keraslayer method I have to give my output shape same as the USE delivers for after embedding a sentence which is TensorShape([1, 512]), right? Like below:

model = "USE_model"
hub_layer = hub.KerasLayer(model, output_shape=[512], input_shape=[], 
                           dtype=tf.string, trainable=True)

**Comment by djdongjin** [comment]
I think you don't need to provide the output_shape. If you use the `hub.KerasLayer` inside a `tf.keras.Sequential`, the output shape of each layer will be computed automatically. But you definitely can play around to see if it is the case.

**Comment by SyedAgha** [comment]
Thanks a lot JIn. I will try and get back in case any problem occurs.

**Comment by anirbankonar123** [comment]
How we can fine-tune the embeddings for my text corpus, it is not clear to me. I dnt have labels as stated here, should be completely Unsupervised.

**Comment by djdongjin** [comment]
I feel most folks here misunderstood the purpose of preptrained embeddings. For any embeddings, e.g., USE, word2vec, BERT, etc, they all aim at providing good initial embeddings to be used in downstream tasks such as classification or translation, etc. That's where fine-tuning happens. Whether the fine-tuning is supervised or unsupervised depends on your downstream task.

I don't have a downstream task, can I still fine-tune the embeddings themselves? Yes you can, but this is not the main purpose of these pretrained embeddings. And the fine-tuning under this line is basically the same as training an embedding from scratch (which indeed is unsupervised), except that now you have a better checkpoint which is trained on a larger corpus. E.g. you need a complete USE or word2vec model and their training methodology.

Back to your question, I mentioned how one may possibly train a USE embedding here.

> For instance, if you want to train a USE model, you need to implement the encoder model, which is a transformer here, and one or more output layers to compute losses (e.g. a binary cross-entropy loss). Then you train the complete model by constructing training data from text.
> 
> With the checkpoint provided by TF Hub (e.g. hub.KerasLayer(...)), you don't need to implement the encoder itself because this is what the magic TH Hub link provides. Then you can start with the checkpoint and add the corresponding output layer and fine-tune the model on your own domain-specific data following the way how they construct training data and compute the loss.
> 
> This is just the way that I think is viable to unsupervised fine-tune a text encoder.

**Comment by djdongjin** [comment]
Also, for me, the purpose of TF-Hub is not to provide a complete pipeline with which you can simply feed the data and get the fine-tuned model (though it is making efforts towards this direction via tools which you can find in the repo).

Instead, the main purpose is to provide pretrained reusable **layers** that can be used in your own models. For instance, you don't need to implement a ResNet by yourself but can simply add it to your model via a TF-Hub layer. That being said, you have to define other parts of your model AND the loss, either supervised or unsupervised, even for a simple binary classification task.

**Comment by AmmarRashed** [comment]
Thanks. Though the fine-tuning I think we are asking about here and in issue [#597 ](https://github.com/tensorflow/hub/issues/597) is the fine-tuning of the language model itself. This is not an unorthodox concept. Flair for example provides that [here](https://github.com/flairNLP/flair/blob/master/resources/docs/TUTORIAL_9_TRAINING_LM_EMBEDDINGS.md). A couple of applications for example, it would be great for example to see how Universal Sentence Encoders family fine-tuned on different domains/cultures cluster certain sentences together, thus exposing the divergence in association rules btw cultures or domains. Another more common example in industry is utterances in Chat bots. Different chat bots for different usages require the model to be fine-tuned on the bot's domain. So when someone says "I want Apple" in tech business the chat bot would not give him oranges.

Thanks a lot for your support, and has been great using TF-hub!

**Comment by djdongjin** [comment]
@AmmarRashed I am not meant to say that fine-tuning language modelling is uncommon. What I said is that TF-Hub simply provides layer-wise components instead of complete pipelines. Then, either a supervised classification or unsupervised language modelling, a complete pipeline including loss and training method is needed.

Maybe it would be better to have a colab tutorial, similar to fine-tuning on sentence classification, to show how to fine-tune language models such as USE :).

**Comment by AmmarRashed** [comment]
That would greatly helpful. Thanks!

**Comment by gowthamkpr** [comment]
@SyedAgha @AmmarRashed Can you guys open a new issue requesting the example. It would be easier to track it that way. Thanks!

**Comment by AmmarRashed** [comment]
@gowthamkpr  Done :) Thanks.

**Comment by gowthamkpr** [comment]
Closing this issue as it has been address. @AmmarRashed Thanks for opening a new issue
